### ISSUE TITLE
long polling, server push, async, AsyncServlet, memory overhead, max handling capacity per jetty instance (per 600mhz cpu & 256mb ram), async http vs. websockets

### ISSUE THREAD

**Issue by r-jo** [eclipse/jetty.project#6492]
**Jetty version**
10.0.5
**Java version**
OpenJDK Runtime Environment (build 11.0.11+9-Ubuntu-0ubuntu2.20.04)

**Question**
I am amateur as far as servers are concerned. I have a google app engine project where some nginx Google Frontend load balancer forwards to my **java 11 web app where I use embedded jetty** (fat jar created with maven shade plugin). I realize I have a lot of responsibilities since if I do something not knowledgeble, I might end up with 10x more app engine instances and bill.

My goal is if I have lets say 10k users 3h a day evenly distributed, I guess 1k users normally concurrently using my web app, one app engine instance (~**1x 600mhz cpu core, 256mb memory**) should be able to handle the load and scale accordingly. I know, it depends.

Web sockets are not possible on app engine standard so if clients communicate with each other I guess best is long polling. Using sync request handling, I see some 100 requests in a very short amount of time that jetty allows if I test it with curl. I guess it is a max number in some queue and I should not tweak it.

However, sync handling should be rather compatible with short polling, long polling would need plenty of clients waiting. Jetty had this **Continuation framework** even before AsyncServlet and I started to get familiar with it, but when I upgraded to 10.0.5 I realized it is **deprecated and I should use AsyncServlet** myself.

I experimented with it and if I want a client to wait for a reply I use:
final AsyncContext asyncContext = request.startAsync(request, response);
and then I store the asyncContext object in a static map (for now, later in some memorystore like memcache or redis).

Jetty can handle **a few 1000 waiting clients** now even if they come in a very short time (concurrent curl requests from bash script). It has some problems around from 3000 requests. What I see is **a memory consumption of about 110mb per 1000 async handled clients**. My questions are:
- am I right that the way to go with client communication (server push, long polling) via jetty 10 is AsyncServlet?
- **how many long polling waiting clients per jetty instance should I aim to handle**? 1000, 2000 or even more is possible? **From 1x 600mhz cpu and 256mb memory**.
- **is it possible to get less memory overhead?** It does not come from the stored static map but from jetty internals I guess, initiated by: request.startAsync(request, response); I guess keeping the connection, request, response objects etc...

I am actually happy if one jetty instance handles 1000 long polling waiting clients with 110mb memory and if needed, app engine starts another instance. I am just curiouos if I am on the right path here and whether it is unevitable to use less memory per async.

I guess it is a complex mechanism in the background with thread pools and one thread handling many async requests (releasing handling thread and keeping connection, plenty of objects in memory) instead of one thread per sync request handling (if I understand it well). So bottlenecks could be at a lot of places I just do not see. So keeping around 1000 waiting async requests and 1 sync & 1 async req handling / sec per 600mhz & 256mb ram may be even overambitious? 

If this is the way to go and 1000 from 110mb memory is what I can expect (which anyways seems to be awesome) I am okay with that. If it somehow could be 3x, 10x that from the same memory, of course it would be even better and cheaper. I would be also good to know if I should aim at less for some other reason.

Another thing is, since on other cloud services (even at google) web sockets are available, I might consider to use them. But it would be some time to become familiar with it and I am not sure if I could handle more communicating clients with them. There could be some general bottlenecks somewhere I do not see (number of connections to hold simultaneously, memory usage, negative effects on normal sync request handling...) Would it be better with web sockets? Could I drain more from one jetty instance or are there some deeper lying bottlenecks so http and websockets are rather equal in this sense?

I hope I did not ramble too much. I am happy with any insights in this topic. Thanks and keep up the good work!

**Comment by r-jo** [comment]
I used this servlet for testing:

package fun.gomo.servlets;

import javax.servlet.AsyncContext;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.logging.Level;
import java.util.logging.Logger;

public class LongPollTest extends HttpServlet {
    private static final Logger logger = Logger.getLogger(fun.gomo.servlets.Login.class.getName());
    private static final long TEN_MINUTES = 10 * 60 * 1000;
    private static final Map<String, AsyncContext> allClientsWaiting = new HashMap<>();

    static {
        logger.setLevel(Level.FINEST);
    }

    protected void doPost(HttpServletRequest request, HttpServletResponse response) {
        logMemoryUsage();
        String talkId = String.valueOf(UUID.randomUUID());
        startWaitingForMessage(request, response, talkId);
        logger.log(Level.INFO, "clients waiting for message: {0}", allClientsWaiting.size());
    }

    private void startWaitingForMessage(HttpServletRequest request, HttpServletResponse response, String talkId) {
        logger.log(Level.INFO, "start waiting...");

        final AsyncContext asyncContext = request.startAsync(request, response);
        asyncContext.setTimeout(TEN_MINUTES);
        allClientsWaiting.put(talkId, asyncContext);
    }

    private void logMemoryUsage() {
        System.gc();
        logger.log(Level.INFO, "used memory: {0} mb", (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1000000);
    }
}

via:
#!/bin/bash
for i in {1..1000}
do
curl -X POST http://localhost:8080/long-poll-test &
done

**Comment by sbordet** [comment]
I strongly suggest that you look into [CometD](https://cometd.org), a sister project of Jetty. It does exactly what you're trying to do.

I just ran the [CometD benchmark](https://docs.cometd.org/current6/reference/#_benchmarking) and after a full GC the server, for 1000 connected clients using HTTP long poll, shows:

```
[2021-07-03T13:30:42.758+0200][info][gc] GC(16) Pause Full (System.gc()) 233M->179M(2048M) 33,322ms
```

So 1000 HTTP clients require 179 MiB.
WebSocket does not change much, about 186 MiB:

```
[2021-07-03T13:40:51.056+0200][info][gc] GC(896) Pause Full (System.gc()) 187M->186M(256M) 22,938ms
```

WebSocket will consume much less CPU and give you smaller latencies.

With the server limited at 256 MiB heap (but a more powerful processor that my laptop has), the CometD benchmark runs easily at about 45 k HTTP responses/s for about 100 k messages/s.

The GC on the server runs back-to-back, but 256 MiB is _really_ small -- I think even fridges and washing machines have more these days 😄 

In summary: 

* Don't write your own solution, use CometD instead.
* Use the CometD benchmark for your tests, setting it up with your expected number of clients and message rate, and see what numbers you get.

Feel free to close this issue and open an issue at the CometD project if you decide to use CometD.

**Comment by r-jo** [comment]
Thanks for your answer! Super useful!
I will take a look at CometD! Thanks again!
F1 instance class is the smallest. You can have other instances as well but they cost linearly more :)
https://cloud.google.com/appengine/docs/standard#second-gen-runtimes
It was 128mb before java11... I guess for sync request processing under small load it is ok.
It has the advantage though that we really have to think about code efficiency :)
